### PR TITLE
[Model] VLM2Vec, the first multimodal embedding model in vLLM

### PR DESCRIPTION
**Pull request by DarkLight1337**
Support [VLM2Vec](https://huggingface.co/TIGER-Lab/VLM2Vec-Full) embedding model by TIGER-Lab.

This is a low-hanging fruit as the model architecture is exactly the same as Phi3V.

Future works, in order of priority:
1. Add CLI option to specify whether to use a model for generation or embedding so we don't have to hardcode the model name when the same model architecture can be used for both.
2. Add multimodal embedding API for OpenAI-compatible server
3. More multimodal embedding models, e.g. [E5-V](https://huggingface.co/royokong/e5-v) which can be similarly supported with our existing LLaVA-NeXT implementation

@jeejeelee are you available to help provide LoRA support for this model after this PR?

### CODE REVIEW

**Comment by jeejeelee** [comment]
Okay, I'd be happy to try supporting LoRA.

File: vllm/model_executor/models/gemma2.py [hunk 1]
```diff
@@ -461,3 +463,50 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
                            if self.config.tie_word_embeddings else None),
         )
         loader.load_weights(weights)
+
+
+class Gemma2EmbeddingModel(nn.Module, SupportsPP):
```
Comment (DarkLight1337) [inline vllm/model_executor/models/gemma2.py:468]: I'm preemptively moving them into the same file to be consistent with the upcoming BERT PR. (#9056)

**Comment by wenhuchen** [comment]
Nice work! There were some typos in our paper. We actually used last token instead eos token representation. I saw that you already used the last token as the representation, which is the correct implementation.

**Review by Isotr0py — APPROVED** [review]
LGTM. Especially the implementation correctness is checked by model vendors :)

**Comment by DarkLight1337** [comment]
> Ooops, seems that the new added vision embedding test didn't include in test-pipeline:
> 
> https://github.com/vllm-project/vllm/blob/7abba39ee64c1e2c84f48d7c38b2cd1c24bb0ebb/.buildkite/test-pipeline.yaml#L337-L349
> 
> We might need to open another PR to include it.

Nice catch, I have opened #9406

**Comment by jvlinsta** [comment]
What will it take / when would it be expected for online engine support for this? ^^

**Comment by DarkLight1337** [comment]
> What will it take / when would it be expected for online engine support for this? ^^

Maybe sometime in the next two weeks.

**Comment by DarkLight1337** [comment]
> > What will it take / when would it be expected for online engine support for this? ^^
> 
> Maybe sometime in the next two weeks.

Quick heads-up that it's done now!

**Comment by wenhuchen** [comment]
When will the LoRA version be online? Actually, the LoRA version works better.

**Comment by DarkLight1337** [comment]
> When will the LoRA version be online? Actually, the LoRA version works better.

Currently, vLLM only supports LoRA for the language backbone of VLMs - some re-arch work is necessary to extend this to the vision encoder. @jeejeelee do you have a timeline regarding this?

**Comment by wenhuchen** [comment]
> > When will the LoRA version be online? Actually, the LoRA version works better.
> 
> Currently, vLLM only supports LoRA for the language backbone of VLMs - some re-arch work is necessary to extend this to the vision encoder. @jeejeelee do you have a timeframe regarding this?

I guess you meant "vLLM doesn't support LoRA" instead of "vLLM only support LoRA"?

**Comment by DarkLight1337** [comment]
> I guess you meant "vLLM doesn't support LoRA" instead of "vLLM only support LoRA"?

I mean that vLLM supports LoRA on the language backbone, but not LoRA on the vision encoder of VLMs.

**Pull request merged**